keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: pod-certificates-example
+spec:
+  restartPolicy: OnFailure
+  automountServiceAccountToken: false
+  containers:
+  - name: main
+    image: debian
+    command: ['sleep', 'infinity']
+    volumeMounts:
+    - name: kube-apiserver-client-certificate
+      mountPath: /var/run/secrets/kubernetes.io/api-mtls-access
+  volumes:
+  - name: kube-apiserver-client-certificate
+    projected:
+      sources:
+      - podCertificate:
+          signerName: "kubernetes.io/kube-apiserver-client-pod"
+          credentialBundlePath: client-credentials.pem
+      - configMap:
+          localObjectReference: kube-root-ca.crt
+          items:
+          - key: ca.crt
+            path: kube-apiserver-root-certificate.pem
+      - downwardAPI:
+          items:
+          - path: namespace
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+```
+
+I ensure that my pod uses client-go's InClusterConfig, and client-go automatically uses the mTLS client certificate to authenticate to kube-apiserver.
+
+### Risks and Mitigations
+
+**Scalability**: In the current design and draft implementation, Kubelet holds
+the private keys for each pod on the node in memory.  If Kubelet restarts, then
+the private keys are lost, and Kubelet will need to immediately generate new
+keys and request new certificates for each volume on the pod.  If Kubelet
+crashloops, this can generate high CertificateSigningRequest traffic on
+kube-apiserver.  This can potentially be mitigated by giving Kubelet the
+capability to read back the existing key from the projected volume on startup.
+
+## Design Details
+
+### X.509 PodIdentity Extension
+
+The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
+the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
+identity information into an X.509 certificate or certificate request, we can
+define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
+define a management policy for the PEN) and specify that it contains the
+following DER-encoded structure:
+
+```go
+type podIdentityASN1 struct {
+	Namespace          string `asn1:"utf8"`
+	ServiceAccountName string `asn1:"utf8"`
+	PodName            string `asn1:"utf8"`
+	PodUID             string `asn1:"utf8"`
+	NodeName           string `asn1:"utf8"`
+}
+```
+
+Certificates and CertificateRequests that embed this extension do not need to include any data in the standard X.509 Name or Subject Alternate Name fields.  The CertificateRequests generated by Kubelet will not, nor will the certificates issued by `kubernetes.io/kube-apiserver-client-pod`.
+
+Utility functions for embedding and extracting this extension from X.509
+Certificate and CertificateRequest objects are defined in
```

Comment:
If talking about objects from a particular Go library, name the library (eg `crypto/x509` from `std`). That makes the context clear to all readers.

---

Alternatively, if talking about data defined by X.509 itself:

```suggestion
Certificate and CertReqMessage objects are defined in
```
(I assume it's CertReqMessage and not CertReq, but haven't checked)

---

Reply:
I was talking about about crypto/x509.  Clarified.